src/bin/casr-cli.rs

```diff
@@ -629,3 +660,217 @@ fn change_text_view(layout1: &mut LinearLayout, act: Action) -> Option<EventResu
+                let mut input = report.canonicalize().unwrap().with_extension("");
+                // input == .*/crash.gdb
+                if input.extension().is_some() && input.extension().unwrap() == "gdb" {
+                    input.set_extension("");
+                }
+                // input == .*/crash
+                if !filestems.insert(input.clone()) {
+                    continue;
+                }
+
+                let mut result: Vec<String> = Vec::new();
+                let crash = if input.exists() {
+                    input.clone()
+                } else {
+                    report
+                }
+                .to_str()
+                .unwrap()
+                .to_string();
+
+                let mut report = input.to_str().unwrap().to_string();
+                report.push_str(".casrep");
+
+                let (san_desc, san_line) = if let Some((report_sum, san_desc, san_line)) =
+                    process_report(&report, "casrep")
+                {
+                    result.push(report_sum);
+                    (san_desc, san_line)
+                } else {
+                    (String::new(), String::new())
+                };
+
+                let report = report.replace(".casrep", ".gdb.casrep");
+                let (casr_gdb_desc, casr_gdb_line) =
+                    if let Some((report_sum, casr_gdb_desc, casr_gdb_line)) =
+                        process_report(&report, "gdb.casrep")
+                    {
+                        result.push(report_sum);
+                        (casr_gdb_desc, casr_gdb_line)
+                    } else {
+                        (String::new(), String::new())
+                    };
+
+                if result.is_empty() {
+                    warn!("Skipping {}. Cannot read casrep.", crash);
+                    continue;
+                } else {
+                    result.push(crash);
+                }
+
+                let mut hash = String::new();
+                hash.push_str(san_desc.as_str());
+                hash.push_str(san_line.as_str());
+                hash.push_str(casr_gdb_desc.as_str());
+                hash.push_str(casr_gdb_line.as_str());
+
+                let san_cnt = cluster_classes.get(&san_desc).unwrap_or(&0) + 1;
+                cluster_classes.insert(san_desc, san_cnt);
+                if !casr_gdb_desc.is_empty() {
+                    let casr_gdb_cnt = cluster_classes.get(&casr_gdb_desc).unwrap_or(&0) + 1;
+                    cluster_classes.insert(casr_gdb_desc, casr_gdb_cnt);
+                }
+
+                let value = if let Some(res) = cluster_hash.get(&hash) {
+                    (res.0.clone(), res.1 + 1)
+                } else {
+                    (result, 1)
+                };
+                cluster_hash.insert(hash, value);
+            }
+        }
+        for info in cluster_hash.values() {
+            info!("{}: {}", "Crash".green(), info.0.last().unwrap());
```

Comment:
Maybe some comments here? for each log line?

Reply:
done